File: experiment-14/runs/language=typescript_model=opus-4.8-fast_prompt=neutral/rep1/src/format.ts
/**
 * Brazilian Soccer MCP — Response formatting
 * ------------------------------------------
 * Context: Turns the structured results from `SoccerStore` into the
 * human-readable text blocks shown in the specification's "Example answer
 * format" sections. Keeping formatting separate from querying means the store
 * stays purely analytical and these helpers stay purely presentational, so both
 * are easy to test in isolation.
 */

import type { Match, Player } from "./types.js";
import type { StandingRow, TeamRecord } from "./store.js";

const pct = (x: number) => `${(x * 100).toFixed(1)}%`;

function score(m: Match): string {
  if (m.homeGoals === null || m.awayGoals === null) return "vs";
  return `${m.homeGoals}-${m.awayGoals}`;
}

function context(m: Match): string {
  const bits: string[] = [m.competition];
  if (m.round) bits.push(`Round ${m.round}`);
  if (m.stage) bits.push(m.stage);
  return bits.join(", ");
}

/** One match as a single line: "2023-09-03: Flamengo 2-1 Fluminense (Brasileirão …)". */
export function formatMatchLine(m: Match): string {
  const date = m.date ?? "date unknown";
  return `${date}: ${m.homeTeam} ${score(m)} ${m.awayTeam} (${context(m)})`;
}

export function formatMatches(matches: Match[], title: string, limit = 25): string {
  if (matches.length === 0) return `${title}\nNo matches found.`;
  const shown = matches.slice(0, limit).map((m) => `- ${formatMatchLine(m)}`);
  const more =
    matches.length > limit ? `\n... (${matches.length - limit} more matches)` : "";
  return `${title} (${matches.length} match${matches.length === 1 ? "" : "es"}):\n${shown.join("\n")}${more}`;
}

export function formatHeadToHead(h: {
  teamA: { display: string } | null;
  teamB: { display: string } | null;
  matches: Match[];
  aWins: number;
  bWins: number;
  draws: number;
  aGoals: number;
  bGoals: number;
}): string {
  if (!h.teamA || !h.teamB) {
    const missing = !h.teamA ? "first" : "second";
    return `Could not resolve the ${missing} team name.`;
  }
  const header = `${h.teamA.display} vs ${h.teamB.display} — head-to-head`;
  const summary =
    `Played: ${h.matches.length} | ${h.teamA.display} ${h.aWins} wins, ` +
    `${h.teamB.display} ${h.bWins} wins, ${h.draws} draws | ` +
    `Goals: ${h.teamA.display} ${h.aGoals}, ${h.teamB.display} ${h.bGoals}`;
  const recent = h.matches.slice(-10).reverse().map((m) => `- ${formatMatchLine(m)}`);
  const body = recent.length ? `\nMost recent meetings:\n${recent.join("\n")}` : "";
  return `${header}\n${summary}${body}`;
}

export function formatTeamRecord(rec: TeamRecord, scope: string): string {
  return (
    `${rec.team} record (${scope}):\n` +
    `- Matches: ${rec.played}\n` +
    `- Wins: ${rec.wins}, Draws: ${rec.draws}, Losses: ${rec.losses}\n` +
    `- Goals For: ${rec.goalsFor}, Goals Against: ${rec.goalsAgainst} ` +
    `(diff ${rec.goalsFor - rec.goalsAgainst >= 0 ? "+" : ""}${rec.goalsFor - rec.goalsAgainst})\n` +
    `- Points: ${rec.points}\n` +
    `- Win rate: ${pct(rec.winRate)}`
  );
}

export function formatStandings(rows: StandingRow[], title: string, limit = 30): string {
  if (rows.length === 0) return `${title}\nNo data available for that competition/season.`;
  const lines = rows.slice(0, limit).map((r) => {
    const champ = r.rank === 1 ? " — Champion" : "";
    return (
      `${String(r.rank).padStart(2)}. ${r.team} — ${r.points} pts ` +
      `(${r.wins}W ${r.draws}D ${r.losses}L, GF ${r.goalsFor} GA ${r.goalsAgainst}, ` +
      `GD ${r.goalDiff >= 0 ? "+" : ""}${r.goalDiff})${champ}`
    );
  });
  return `${title}:\n${lines.join("\n")}`;
}

export function formatCompetitionStats(
  stats: {
    matches: number;
    decided: number;
    totalGoals: number;
    avgGoals: number;
    homeWins: number;
    awayWins: number;
    draws: number;
    homeWinRate: number;
    biggestMargins: Match[];
  },
  title: string,
): string {
  if (stats.decided === 0) return `${title}\nNo decided matches found for that scope.`;
  const big = stats.biggestMargins
    .slice(0, 5)
    .map((m, i) => `${i + 1}. ${formatMatchLine(m)}`)
    .join("\n");
  return (
    `${title}:\n` +
    `- Matches: ${stats.matches} (${stats.decided} with recorded scores)\n` +
    `- Total goals: ${stats.totalGoals}\n` +
    `- Average goals per match: ${stats.avgGoals.toFixed(2)}\n` +
    `- Home wins: ${stats.homeWins} (${pct(stats.homeWinRate)}), ` +
    `Away wins: ${stats.awayWins}, Draws: ${stats.draws}\n` +
    `\nBiggest victories:\n${big}`
  );
}

export function formatPlayerLine(p: Player, idx?: number): string {
  const prefix = idx !== undefined ? `${idx + 1}. ` : "";
  const rating = p.overall !== null ? `Overall: ${p.overall}` : "Overall: ?";
  const pos = p.position || "?";
  const club = p.club || "Free agent";
  return `${prefix}${p.name} — ${rating}, Position: ${pos}, Club: ${club} (${p.nationality}, age ${p.age ?? "?"})`;
}

export function formatPlayers(players: Player[], title: string, limit = 20): string {
  if (players.length === 0) return `${title}\nNo players found.`;
  const lines = players.slice(0, limit).map((p, i) => formatPlayerLine(p, i));
  const more = players.length > limit ? `\n... (${players.length - limit} more players)` : "";
  return `${title} (${players.length} found):\n${lines.join("\n")}${more}`;
}

File: experiments/adrianco/experiment-2/runs/language=go_model=opus_tooling=beads/rep1/internal/mcp/format.go
package mcp

import (
	"fmt"
	"strings"

	"brsoccer/internal/data"
	"brsoccer/internal/query"
)

func FormatMatches(ms []data.Match, team, vs string) string {
	if len(ms) == 0 {
		return "No matches found."
	}
	var sb strings.Builder
	if team != "" && vs != "" {
		sb.WriteString(fmt.Sprintf("%s vs %s — %d matches:\n", team, vs, len(ms)))
	} else {
		sb.WriteString(fmt.Sprintf("%d matches:\n", len(ms)))
	}
	for _, m := range ms {
		date := "????-??-??"
		if !m.Date.IsZero() {
			date = m.Date.Format("2006-01-02")
		}
		extra := m.Competition
		if m.Round != "" {
			extra += " Round " + m.Round
		}
		if m.Stage != "" {
			extra += " (" + m.Stage + ")"
		}
		sb.WriteString(fmt.Sprintf("  %s: %s %d-%d %s [%s]\n",
			date, m.HomeTeam, m.HomeGoals, m.AwayGoals, m.AwayTeam, extra))
	}
	return sb.String()
}

func FormatTeamStats(s query.TeamStats) string {
	var sb strings.Builder
	header := s.Team
	if s.Season != 0 {
		header += fmt.Sprintf(" (%d)", s.Season)
	}
	if s.Competition != "" {
		header += " — " + s.Competition
	}
	sb.WriteString(header + "\n")
	sb.WriteString(fmt.Sprintf("  Matches: %d\n", s.Matches))
	sb.WriteString(fmt.Sprintf("  W-D-L: %d-%d-%d (%.1f%% win rate)\n", s.Wins, s.Draws, s.Losses, s.WinRate()))
	sb.WriteString(fmt.Sprintf("  Goals For / Against: %d / %d (diff %+d)\n", s.GoalsFor, s.GoalsAgainst, s.GoalsFor-s.GoalsAgainst))
	sb.WriteString(fmt.Sprintf("  Home W-D-L: %d-%d-%d\n", s.HomeWins, s.HomeDraws, s.HomeLosses))
	sb.WriteString(fmt.Sprintf("  Away W-D-L: %d-%d-%d\n", s.AwayWins, s.AwayDraws, s.AwayLosses))
	sb.WriteString(fmt.Sprintf("  Points (3W+D): %d\n", s.Points))
	return sb.String()
}

func FormatH2H(h query.HeadToHead) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("%s vs %s head-to-head:\n", h.TeamA, h.TeamB))
	sb.WriteString(fmt.Sprintf("  Matches: %d\n", h.Matches))
	sb.WriteString(fmt.Sprintf("  %s wins: %d, %s wins: %d, Draws: %d\n", h.TeamA, h.AWins, h.TeamB, h.BWins, h.Draws))
	sb.WriteString(fmt.Sprintf("  Goals: %s %d - %d %s\n", h.TeamA, h.AGoals, h.BGoals, h.TeamB))
	if len(h.Examples) > 0 {
		sb.WriteString("Recent meetings:\n")
		sb.WriteString(FormatMatches(h.Examples, "", ""))
	}
	return sb.String()
}

func FormatStandings(st []query.TeamStats, season int, competition string) string {
	if len(st) == 0 {
		return "No data for this season/competition."
	}
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("%s %d standings (computed):\n", competition, season))
	sb.WriteString(fmt.Sprintf("  %-3s %-30s %3s %3s %3s %3s %4s %4s %4s\n", "#", "Team", "P", "W", "D", "L", "GF", "GA", "Pts"))
	for i, s := range st {
		sb.WriteString(fmt.Sprintf("  %-3d %-30s %3d %3d %3d %3d %4d %4d %4d\n",
			i+1, trunc(s.Team, 30), s.Matches, s.Wins, s.Draws, s.Losses, s.GoalsFor, s.GoalsAgainst, s.Points))
	}
	return sb.String()
}

func FormatPlayers(ps []data.Player) string {
	if len(ps) == 0 {
		return "No players found."
	}
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("%d players:\n", len(ps)))
	for _, p := range ps {
		sb.WriteString(fmt.Sprintf("  %s — Overall %d, %s, %s @ %s\n",
			p.Name, p.Overall, p.Position, p.Nationality, p.Club))
	}
	return sb.String()
}

func FormatOverall(g query.GlobalStats, comp string, season int) string {
	var sb strings.Builder
	header := "Dataset statistics"
	if comp != "" {
		header += " — " + comp
	}
	if season != 0 {
		header += fmt.Sprintf(" %d", season)
	}
	sb.WriteString(header + "\n")
	sb.WriteString(fmt.Sprintf("  Matches: %d\n", g.TotalMatches))
	sb.WriteString(fmt.Sprintf("  Total goals: %d\n", g.TotalGoals))
	sb.WriteString(fmt.Sprintf("  Average goals/match: %.2f\n", g.AvgGoalsPerMatch))
	sb.WriteString(fmt.Sprintf("  Home wins: %d (%.1f%%)\n", g.HomeWins, g.HomeWinRate))
	sb.WriteString(fmt.Sprintf("  Away wins: %d\n", g.AwayWins))
	sb.WriteString(fmt.Sprintf("  Draws: %d\n", g.Draws))
	return sb.String()
}

func trunc(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n-1]) + "…"
}
